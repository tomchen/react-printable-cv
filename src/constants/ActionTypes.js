export const CHANGE_LANG = 'CHANGE_LANG'

export const RENDER_DATA = 'RENDER_DATA'

export const IMPORT_JSON = 'IMPORT_JSON'
