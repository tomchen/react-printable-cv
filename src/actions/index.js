import i18n from 'i18next'
import * as types from '../constants/ActionTypes'
import { userData, projectData, loadedLocale } from '../utils/dataCache'

const getDataLang = async (lang, userOrProject) => {
  // userOrProject: true=user ; false=project
  if (lang in (userOrProject ? userData : projectData)) {
    return userData.lang
  }

  const d = await import(
    `Data/${userOrProject ? 'cv' : 'project-list'}/${lang}.json`
  )
  return d
}

const getI18nLang = async (lang) => {
  if (loadedLocale.includes(lang)) {
    return true
  }
  const res = await import(`../../i18n/${lang}/common.json`)
  i18n.addResourceBundle(lang, 'common', res)
  loadedLocale.push(lang)
  return true
}

export const renderData = (data) => ({ type: types.RENDER_DATA, data })

export const changeLang = (lang) => ({ type: types.CHANGE_LANG, lang })

export const changeLangAsync = (lang) => (dispatch) => {
  Promise.all([
    getDataLang(lang, true),
    getDataLang(lang, false),
    getI18nLang(lang),
  ]).then(([fetchedUserData, fetchedProjectData]) => {
    dispatch(changeLang(lang))
    dispatch(
      renderData({
        userData: fetchedUserData,
        projectData: fetchedProjectData,
      }),
    )
  })
}

export const importJson = (data) => ({ type: types.IMPORT_JSON, data })
