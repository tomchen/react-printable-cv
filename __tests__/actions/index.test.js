import configureStore from 'redux-mock-store'
import * as actions from 'Src/actions'
import * as types from 'Src/constants/ActionTypes'

const mockStore = configureStore()
const store = mockStore()

describe('test actions', () => {
  beforeEach(() => {
    store.clearActions()
  })

  it('changeLang dispatches the correct action and payload', () => {
    const lang = 'zh-cn'
    store.dispatch(actions.changeLang(lang))
    const expectedActions = [
      {
        lang,
        type: types.CHANGE_LANG,
      },
    ]
    const s = expect(store.getActions())
    s.toEqual(expectedActions)
    s.toMatchSnapshot()
  })

  it('renderData dispatches the correct action and payload', () => {
    const data = { a: 6 }
    store.dispatch(actions.renderData(data))
    const expectedActions = [
      {
        data,
        type: types.RENDER_DATA,
      },
    ]
    const s = expect(store.getActions())
    s.toEqual(expectedActions)
    s.toMatchSnapshot()
  })

  it('importJson dispatches the correct action and payload', () => {
    const data = { b: 'g' }
    store.dispatch(actions.importJson(data))
    const expectedActions = [
      {
        data,
        type: types.IMPORT_JSON,
      },
    ]
    const s = expect(store.getActions())
    s.toEqual(expectedActions)
    s.toMatchSnapshot()
  })
})
