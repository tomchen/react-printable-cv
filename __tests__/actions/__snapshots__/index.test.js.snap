// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test actions changeLang dispatches the correct action and payload 1`] = `
Array [
  Object {
    "lang": "zh-cn",
    "type": "CHANGE_LANG",
  },
]
`;

exports[`test actions importJson dispatches the correct action and payload 1`] = `
Array [
  Object {
    "data": Object {
      "b": "g",
    },
    "type": "IMPORT_JSON",
  },
]
`;

exports[`test actions renderData dispatches the correct action and payload 1`] = `
Array [
  Object {
    "data": Object {
      "a": 6,
    },
    "type": "RENDER_DATA",
  },
]
`;
